File: matlab/P2L3Err.m
function [resid] = P2L3Err(parms,fdv)

[EV,IV,vols] = P2L3(parms,fdv);
tot=(IV(1:2,:)*vols(1))+(EV(1:2,:)*vols(2))+(EV(3:4,:)*vols(3));
resid=(fdv.data-tot);
%Normalize to max of measured data. May need to re-weight...
resid(1,:)=resid(1,:)/max(fdv.data(1,:));
resid(2,:)=resid(2,:)/max(fdv.data(2,:));

resid=reshape(resid,1,prod(size(fdv.data))); 

File: python/P2L3Err.py
import numpy as np
import P2L3

def P2L3ErrOuter(fdv):
    '''
    Function to calculate the residuals for use with scipy least_squares() func

    Input:
        fdv:        For information about the keys in this dictionary, call help(P2L3)

    Output:
        P2L1Inner:   A function of only vars that plays nice with least_squares
    '''

    def P2L3ErrInner(parms):
            '''
            Function of one variable to calculate the residuals for use with scipy least_squares() func

            Input:
                vars:       A two element array of test parameters

            Output:
                resid:       A 1D array containing residuals 
            '''

            EV, IV, vols, Mzev, Mziv = P2L3.P2L3(parms, fdv)
            tot = IV[0:2]*vols[0] + EV[0:2]*vols[1] + EV[2:4]*vols[2]

            resid = fdv['data'] - tot
            # Normalize to max of measured data. May need to reweight
            resid[0] /= np.max( fdv['data'][0] )
            resid[1] /= np.max( fdv['data'][1] )

            resid = resid.reshape(( np.prod(np.shape(resid)), ))
            
            return resid

    return P2L3ErrInner


def P2L3Err(parms, fdv):
    '''
    Function of two variables to calculate the residuals to display or for further processing       
    Input:
        fdv:        For information about the keys in this dictionary, call help(P2L3)
        pamrs:      A two element array of test parameters      
    Output:
        resid:      A 1D array containing residuals
    '''
      
    inner = P2L3ErrOuter(fdv)
    resid = inner(parms)

    return resid
      